br/pkg/task/stream.go

```diff
@@ -941,38 +941,65 @@ func RunStreamTruncate(c context.Context, g glue.Glue, cmdName string, cfg *Stre
 	removed := metas.RemoveDataBefore(shiftUntilTS)
 
 	// remove log
-	clearDataFileDone := console.ShowTask(
-		"Clearing data files... ", glue.WithTimeCost(),
+	p := console.StartProgressBar(
+		"Clearing Data Files", len(removed),
+		glue.WithTimeCost(),
 		glue.WithConstExtraField("kv-count", kvCount),
 		glue.WithConstExtraField("kv-size", fmt.Sprintf("%d(%s)", totalSize, units.HumanSize(float64(totalSize)))),
 	)
+	defer p.Close()
 	worker := utils.NewWorkerPool(128, "delete files")
-	wg := new(sync.WaitGroup)
+	const keepFirstNFailure = 16
+	var notDeleted []string
 	for _, f := range removed {
-		if !cfg.DryRun {
-			wg.Add(1)
-			finalFile := f
-			worker.Apply(func() {
-				defer wg.Done()
-				if err := storage.DeleteFile(ctx, finalFile.Path); err != nil {
-					log.Warn("File not deleted.", zap.String("path", finalFile.Path), logutil.ShortError(err))
-					console.Print("\n"+em(finalFile.Path), "not deleted, you may clear it manually:", warn(err))
-				}
-			})
+		worker.Apply(func() {
+			// NOTE: We cannot get the returned value from deleting, so polling the context here again.
+			//       Maybe we'd better use an `ErrGroup` and fail fast for some types of error (or if len(notDeleted) > N)?
+			//       (It would be a disaster if we have waiting for 6hrs and eventually found out nothing deleted :| )
+			if ctx.Err() != nil {
+				p.Close()
+				return
+			}
+			defer p.Inc()
+			if cfg.DryRun {
+				return
+			}
+			if err := storage.DeleteFile(ctx, f.Path); err != nil {
+				log.Warn("File not deleted.", zap.String("path", f.Path), logutil.ShortError(err))
+				notDeleted = append(notDeleted, f.Path)
+			}
+		})
+	}
+	if err := p.Wait(ctx); err != nil {
```

Comment:
can we wait all worker finish apply here?

Reply:
If the total count is `0`, the progress would finish as soon as it being created.